src/neuroglancer/datasource/computed/backend.ts

```diff
@@ -0,0 +1,594 @@
+    copyFunction(j, k);
+  }
+}
+
+
+
+/**
+ * Computes a consistent key string from a chunk grid position.
+ *
+ * It's tempting to use chunk.key, in particular because these values will
+ * often be the same, but we won't always have access to a fully-specified
+ * chunk, and there's no contractual guarantee that its key will be equal to
+ * the value returned here.
+ * @param gridPosition chunk grid position
+ */
+function gridPositionKey(gridPosition: vec3) {
+  return gridPosition.toLocaleString();
+}
+
+// In addition to acting as a VolumeChunk for the purposes of a ChunkManager
+// object, also performs the book-keeping necessary to prepare the data buffer
+// used as input by the computation that provides its data. This includes
+// fetching chunk data from other datasources.
+export class ComputedVolumeChunk extends VolumeChunk implements ChunkStateListener {
+  // Defines the input and output geometry and datatypes.
+  private computationParams_?: ComputationParameters;
+
+  // Inidicates a cancellation of this chunk's computation.
+  private cancellationToken_?: CancellationToken;
+
+  // Resolve and reject functions correspond to a Promise, returned by getPromise().
+  private resolve_?: () => void;
+  private reject_?: (reason: Error) => void;
+
+  // True iff this chunk is actively computing.
+  private computing_ = false;
+
+  // True iff this chunk has been initialized for computation.
+  private initialized_ = false;
+
+  // A map from grid position string keys, as returned by gridPositionKey to
+  // vec3 grid positions. This is used as an indirection to avoid storing
+  // explicit references to VolumeChunks belonging to the origin source.
+  private originGridPositions_ = new Map<String, vec3>();
+
+  // The lower bound of the input patch.
+  private inputLower_?: vec3;
+
+  // Represents the input patch.
+  private inputBuffer_?: ArrayBuffer;
+
+  // Indicate to the ChunkManager that this is a computational Chunk.
+  isComputational = true;
+
+  // Our source is a ComputedVolumeChunkSource.
+  source: ComputedVolumeChunkSource;
+
+  /**
+   * Sets up computation parameters, computes overlapping origin chunks and
+   * initializes the input buffer. Returns a Promise that will resolve when
+   * computation completes, or reject if computation fails or is cancelled.
+   * @param computationParams computation parameters
+   * @param cancellationToken cancellation token
+   */
+  initializeComputation(
+      computationParams: ComputationParameters, cancellationToken: CancellationToken) {
+    if (!this.source) {
+      throw new Error('initializeComputation must be called after source is valid.');
+    }
+    if (!this.chunkDataSize) {
+      throw new Error('initializeComputation must be called after computeChunkBounds.');
+    }
+    this.computationParams_ = computationParams;
+    this.cancellationToken_ = cancellationToken;
+    this.computing_ = false;
+
+    this.cancellationToken_.add(() => {
+      if (this.reject_) {
```

Comment:
This should also reset this.inputBuffer_ so that if the compute is cancelled, but the chunk remains in QUEUED state because it is still needed, the memory is not held.

Reply:
Done.